File: interpretado/filaListaEncadeada.py
from Node import Node

class FilaListaEncadeada:
    def __init__(self):
        self.front =  None      #front = inicio da fila
        self.rear = None        #rear = final da fila
    def isEmpty(self):
        return (self.rear == None)  #funcao que verifica se a fila esta vazia

    def enqueue(self, valor):
        newNode = Node(valor)  #cria um novo node que sera adicionado na fila
        if self.isEmpty():
            self.rear = self.front = newNode  #define o novo node como o front e o rear, ja que ele sera o primeiro da fila
        else:
            self.rear.next = newNode  #apenas define o node como o novo valor da fila
            self.rear = newNode
    
    def dequeue(self):
        if self.isEmpty():
            print("erro! lista vazia..")
            return None
        valor = self.front.dado  #aloca o dado do ultimo node da fila
        self.front = self.front.next  #define o valor 
        if self.front is None:
            self.rear = None
        return valor
    
    def len(self):
        nodeLocal = self.front
        i = 0

        while nodeLocal is not None:    #percorre a fila - O(n)
            nodeLocal = nodeLocal.next
            i += 1
        return i 

File: compilado/src/filaListaEncadeada.c
#include "../includes/libs.h"
#include "../includes/filaListaEncadeada.h"

FilaEncadeada* criarFilaEncadeada() {
    FilaEncadeada* newFila = (FilaEncadeada*) malloc(sizeof(FilaEncadeada));  //aloca espaço necessario da memoria pra fila
    newFila->inicio = NULL;  //seta os ponteiros do inicio e fim como nulos
    newFila->fim = NULL;  
    return newFila;  //retorna a nova fila
}

bool isEmptyFilaEncadeada(FilaEncadeada* fila) {
    return(fila->fim == NULL);
}

void enqueueFilaEncadeada(FilaEncadeada* fila, int valor) {
    No* newNo = (No*) malloc(sizeof(No));  //aloca memoria para o novo node e prepara ele para inserir na fila
    newNo->dado = valor;
    newNo->proximo = NULL;

    if (isEmptyFilaEncadeada(fila)) {  //caso nao tenha nenhum node no final da fila, ela esta vazia 
        fila->inicio = newNo; //logo o novo node é o inicial e o final da fila
        fila->fim = newNo;

    } else {  //caso onde a fila contem outros itens ja
        fila->fim->proximo = newNo;  //o ultimo elemento aponto para o novo node  
        fila->fim = newNo;  //o final da fila se atualiza
    }
    
}

int dequeueFilaEncadeada(FilaEncadeada* fila) {
    if (isEmptyFilaEncadeada(fila)){
        printf("erro! a fila esta vazia..");  //verifica se a pilha ta vazia
        return -1;
    }
    
    No* noRemovido = fila->inicio;  //aloca o node que sera removido em uma variavel local
    int valorRemovido = noRemovido->dado; // aloca o valor do node removido em uma variavel local

    fila->inicio = noRemovido->proximo; //atualiza o inicio da fila

    if (fila->inicio == NULL)    {  //se a fila ficou vazia, entao o seus valores de incio e fim devem ficar nulos 
        fila->fim = NULL;
    }

    free(noRemovido);  //libera o espaco alocado pelo node
    return valorRemovido;  //retorna o valor do node removido
    
}

int lenFilaEncadeada(FilaEncadeada* fila) {
    int i =  0;
    No* nodeLocal = fila->inicio; //aolca o node no inicio da fila em uma variavel local

    while (nodeLocal != NULL) {  //percorre a fila ate encontrar o elemento nulo, aumentando o contador em cada ocorrencia
        i++;
        nodeLocal = nodeLocal->proximo;
    }
    return i;  //retorna o contador indicando o numero de itens na fila
    
}